src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,285 @@
+        name == "webSocketMessage" ||
+        name == "webSocketClose" ||
+        name == "webSocketError") {
+      return true;
+    }
+    return false;
+  }
+
+  // If the `methodName` is a known public method, we'll return it.
+  inline v8::Local<v8::Function> tryGetFn(
+      Worker::Lock& lock,
+      kj::Maybe<kj::StringPtr> entrypointName,
+      IoContext& ctx,
+      kj::StringPtr methodName) {
+
+    auto& handler = KJ_REQUIRE_NONNULL(lock.getExportedHandler(entrypointName, ctx.getActor()),
+        "Failed to get handler to worker.");
+
+    auto handle = handler.self.getHandle(lock);
+    auto methodStr = jsg::v8StrIntern(lock.getIsolate(), methodName);
+    auto fnHandle = jsg::check(handle->Get(lock.getContext(), methodStr));
+
+    jsg::Lock& js(lock);
+    v8::Local<v8::Object> obj = js.obj();
+    auto objProto = obj->GetPrototype().As<v8::Object>();
+
+    // Get() will check the Object and the prototype chain. We want to verify that the function
+    // we intend to call is not the one defined on the Object prototype.
+    bool isImplemented = fnHandle != jsg::check(objProto->Get(js.v8Context(), methodStr));
+
+    JSG_REQUIRE(isImplemented && fnHandle->IsFunction(), TypeError,
+        kj::str("The RPC receiver does not implement the method \"", methodName, "\"."));
+    JSG_REQUIRE(!isReservedName(methodName), TypeError,
+        kj::str("'", methodName, "' is a reserved method and cannot be called over RPC."));
+    return fnHandle.As<v8::Function>();
   }
+
+  // Deserializes the arguments and passes them to the given function.
+  v8::Local<v8::Value> invokeFn(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+    // We received arguments from the client, deserialize them back to JS.
+    if (serializedArgs.size() > 0) {
+      auto args = KJ_REQUIRE_NONNULL(
+          deserializeV8(js, serializedArgs).tryCast<jsg::JsArray>(),
+          "expected JsArray when deserializing arguments.");
+      // Call() expects a `Local<Value> []`... so we populate an array.
+      KJ_STACK_ARRAY(v8::Local<v8::Value>, arguments, args.size(), 8, 8);
+      for (size_t i = 0; i < args.size(); ++i) {
+        arguments[i] = args.get(js, i);
+      }
+      return jsg::check(fn->Call(js.v8Context(), fn, args.size(), arguments.begin()));
+    } else {
+      return jsg::check(fn->Call(js.v8Context(), fn, 0, nullptr));
+    }
+  };
+
+  struct Returned {
+    kj::Array<kj::byte> serializedValue;
+  };
+  struct ThrewException {
+    kj::Array<kj::byte> serializedException;
+  };
+  // The serialized JS value.
+  using SerializedResult = kj::OneOf<Returned, ThrewException>;
+
+  // Invokes the given function and serializes the result for RPC.
+  jsg::Promise<SerializedResult> callFuncAndSerialize(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+
+    return js.toPromise(invokeFn(js, fn, serializedArgs))
```

Comment:
I think you need to bring back the `js.tryCatch()` around `invokeFn`. Otherwise, if the application code throws an exception, that exception is going to be "tunneled" by converting it to a KJ exception and back. That may lose some information compared to serializing the exception using v8 serialization. It's OK to rely on tunneling for exceptions thrown by the runtime itself, since they don't generally have any extra information on them anyway. But for exceptions thrown by the application, I think we want to use v8 serialization.

The way to test this is something like:

```
let e = new Error("foo");
e.myField = 123;
throw e;
```

Then verify that `myField` is intact when catching on the client side.

(Then again, I could be convinced that we should _never_ serialize exceptions and just always use tunneling. In that case we'd want to change the capnp protocol so that it no longer has explicit v8-serialized exceptions.)

Reply:
Wait... I'm wrong. V8 serialization does not preserve properties added manually to the error object! This is easy to test using the `structuredClone` function in Chrome devtools. Does v8 serialization preserve anything about errors that our existing tunneling doesn't?

I guess v8 serialization will behave "correctly" if the application throws a value that is not an error, e.g. throwing a bare string. People are "not supposed to" do that. Do we care about supporting them? If we don't care, maybe we just shouldn't explicitly catch and serialize exceptions at all. Would certainly simplify the code.

@jasnell thoughts?

Reply:
Another line of argument here might be: We have talked about making regular exception tunneling itself use v8 serialization. If we did that, then the JS RPC protocol's explicit representation of exceptions becomes redundant, we can entirely rely on tunneling.

Given that, I am now leaning towards: We should not explicitly handle exceptions in the JS RPC protocol. We should just throw tunneled exceptions. We can separately improve tunneled exceptions to use v8 serialization if we want to, but that's a separate matter from JS RPC.

Reply:
Right, that's exactly what I meant above. I think this is OK.

This problem already exists today for regular fetch() requests: If the DO's fetch() handler throws the `number` 10, the client will instead catch an `Error` whose description is `10`. I don't see any reason why we should solve this _specifically_ for RPC and not for all use cases. So I think RPC should depend on the same exception tunneling logic as everyone else, and separately later on we can improve that logic.

Reply:
Agree that it's better to be consistent (and it really does simplify the code a lot). I also inlined `handleRpcResponse()` and `callFuncAndSerialize()` at their respective call sites.

https://github.com/cloudflare/workerd/compare/004425cb34c59b5488a066bdcdf72332f49fc8a4..af4e992b21e805f0f31ddcf5f9498ea1a7546429

Reply:
> Does v8 serialization preserve anything about errors that our existing tunneling doesn't?

Missed the ping in this thread. v8 serialization will only serialize the intrinsic properties of an Error, including the stack. Our existing tunneling does not preserve the stack. Using Node.js' structuredClone as a quick way to illustrate (since it's using v8 serialization under the covers)

<img width="548" alt="image" src="https://github.com/cloudflare/workerd/assets/439929/d3398f11-2ee0-485c-925f-3ad3b2daa8f5">

Own properties that are added to the error object are not preserved across the serialization.